File: app/src/main/java/com/rifafauzi/moviecatalogue/di/Injection.java
package com.rifafauzi.moviecatalogue.di;

import android.app.Application;

import com.rifafauzi.moviecatalogue.helper.local.LocalDataSource;
import com.rifafauzi.moviecatalogue.helper.local.room.MoviesDatabase;
import com.rifafauzi.moviecatalogue.helper.remote.ApiClient;
import com.rifafauzi.moviecatalogue.helper.remote.ApiInterface;
import com.rifafauzi.moviecatalogue.helper.remote.RemoteDataSource;
import com.rifafauzi.moviecatalogue.helper.repository.Repository;
import com.rifafauzi.moviecatalogue.utils.AppExecutors;

public class Injection {

    public static Repository repository(Application application) {

        MoviesDatabase database = MoviesDatabase.getInstance(application);

        RemoteDataSource remoteRepository = RemoteDataSource.getInstance(ApiClient.getClient().create(ApiInterface.class));
        LocalDataSource localRepository = LocalDataSource.getInstance(database.moviesDao());
        AppExecutors appExecutors = new AppExecutors();

        return Repository.getInstance(remoteRepository, localRepository, appExecutors);
    }

}

File: app/src/main/java/com/rifafauzi/moviecatalogue/di/Injection.kt
package com.rifafauzi.moviecatalogue.di

import com.rifafauzi.moviecatalogue.helper.remote.ApiClient
import com.rifafauzi.moviecatalogue.helper.remote.ApiInterface
import com.rifafauzi.moviecatalogue.helper.repository.Repository
import com.rifafauzi.moviecatalogue.helper.remote.RemoteDataSource

object Injection {

    fun repository(): Repository? {
        val remoteDataSource = RemoteDataSource.getInstance(ApiClient.client.create(ApiInterface::class.java))
        return Repository.getInstance(remoteDataSource)
    }

}
